### ISSUE TITLE
AddressSanitizer errors in OpenProXL tests

### ISSUE THREAD

**Issue by jpfeuffer** [OpenMS/OpenMS#2696]
Built from my Asan_fix branch. Not sure if completely uptodate but pretty new:
[abibuilder.informatik.uni-tuebingen.de.txt](https://github.com/OpenMS/OpenMS/files/1064604/abibuilder.informatik.uni-tuebingen.de.txt)

**Comment by jpfeuffer** [comment]
Ah maybe it was your fix for the integer overflow. Might not be in there yet.

**Comment by timosachsenberg** [comment]
````
	<Test Status="failed">
		<Name>UTILS_OpenPepXL_1</Name>
		<Path>./src/tests/topp</Path>
		<FullName>./src/tests/topp/UTILS_OpenPepXL_1</FullName>
		<FullCommandLine>/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL &quot;-test&quot; &quot;-in&quot; &quot;/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/source/src/tests/topp/OpenPepXL_input.mzML&quot; &quot;-consensus&quot; &quot;/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/source/src/tests/topp/OpenPepXL_input.consensusXML&quot; &quot;-database&quot; &quot;/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/source/src/tests/topp/OpenPepXL_input.fasta&quot; &quot;-out_xquestxml&quot; &quot;OpenPepXL_output.xquest.xml.tmp&quot; &quot;-out_xquest_specxml&quot; &quot;OpenPepXL_output.spec.xml.tmp&quot; &quot;-out_mzIdentML&quot; &quot;OpenPepXL_output.mzid.tmp&quot; &quot;-out_idXML&quot; &quot;OpenPepXL_output.idXML.tmp&quot;</FullCommandLine>
		<Results>
			<NamedMeasurement type="text/string" name="Exit Code"><Value>OTHER_FAULT</Value></NamedMeasurement>
			<NamedMeasurement type="text/string" name="Exit Value"><Value>1</Value></NamedMeasurement>
			<NamedMeasurement type="numeric/double" name="Execution Time"><Value>3.92163</Value></NamedMeasurement>
			<NamedMeasurement type="text/string" name="Completion Status"><Value>Completed</Value></NamedMeasurement>
			<NamedMeasurement type="text/string" name="Command Line"><Value>/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL &quot;-test&quot; &quot;-in&quot; &quot;/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/source/src/tests/topp/OpenPepXL_input.mzML&quot; &quot;-consensus&quot; &quot;/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/source/src/tests/topp/OpenPepXL_input.consensusXML&quot; &quot;-database&quot; &quot;/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/source/src/tests/topp/OpenPepXL_input.fasta&quot; &quot;-out_xquestxml&quot; &quot;OpenPepXL_output.xquest.xml.tmp&quot; &quot;-out_xquest_specxml&quot; &quot;OpenPepXL_output.spec.xml.tmp&quot; &quot;-out_mzIdentML&quot; &quot;OpenPepXL_output.mzid.tmp&quot; &quot;-out_idXML&quot; &quot;OpenPepXL_output.idXML.tmp&quot;</Value></NamedMeasurement>
			<Measurement>
				<Value>File &apos;/Users/builder/.OpenMS/OpenMS.ini&apos; is deprecated.
Updating missing/wrong entries in &apos;/Users/builder/.OpenMS/OpenMS.ini&apos; with defaults!
XLinks Tolerance: 0.3
Progress of &apos;loading spectra list&apos;:
  22.10 %               -- done [took 0.39 s (CPU), 0.40 s (Wall)] -- 
Progress of &apos;loading chromatogram list&apos;:
-- done [took 0.00 s (CPU), 0.00 s (Wall)] -- 
Progress of &apos;Filtering spectra...&apos;:
-- done [took 0.05 s (CPU), 0.05 s (Wall)] -- 
Progress of &apos;Load database from FASTA file...&apos;:
-- done [took 0.00 s (CPU), 0.00 s (Wall)] -- 
Progress of &apos;Map spectrum precursors to linked features...&apos;:
-- done [took 0.06 s (CPU), 0.06 s (Wall)] -- 
Progress of &apos;Preprocessing Spectra Pairs...&apos;:
-- done [took 0.39 s (CPU), 0.39 s (Wall)] -- 
Progress of &apos;Digesting peptides...&apos;:
-- done [took 0.03 s (CPU), 0.03 s (Wall)] -- 
Number of paired precursor masses: 50
Filtering peptides with precursors
Progress of &apos;Enumerating cross-links...&apos;:
Enumerated pairs with sequence 276 of 276;	 Current pair count: 26 | current size in mb: 0
-- done [took 0.02 s (CPU), 0.02 s (Wall)] -- 
Enumerated cross-links: 26
Sorting of enumerated precursors finished
Progress of &apos;Matching to theoretical spectra and scoring...&apos;:
Processing spectrum pair 1 / 25
Processing spectrum pair 2 / 25
Number of candidates for this spectrum: 0
Processing spectrum pair 3 / 25
Number of candidates for this spectrum: 0
Processing spectrum pair 4 / 25
Number of candidates for this spectrum: 1
=================================================================
==1844==ERROR: AddressSanitizer: container-overflow on address 0x6140000d5ee8 at pc 0x00010e03e0e2 bp 0x7fff51bd8d70 sp 0x7fff51bd8d68
READ of size 8 at 0x6140000d5ee8 thread T0
==1844==atos returned an error: An admin user name and password is required to enter Developer Mode.
    #0 0x10e03e0e1 in TOPPOpenPepXL::main_(int, char const**) (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL+0x10003a0e1)
    #1 0x1122c06e9 in OpenMS::TOPPBase::main(int, char const**) (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/lib/libOpenMS.dylib+0x40fc6e9)
    #2 0x10e007d87 in main (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL+0x100003d87)
    #3 0x7fff950f75ac in start (/usr/lib/system/libdyld.dylib+0x35ac)

0x6140000d5ee8 is located 168 bytes inside of 432-byte region [0x6140000d5e40,0x6140000d5ff0)
allocated by thread T0 here:
    #0 0x114ef65fb in wrap__Znwm (/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/8.0.0/lib/darwin/libclang_rt.asan_osx_dynamic.dylib+0x575fb)
    #1 0x10e12bfde in std::__1::__split_buffer&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch, std::__1::allocator&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch&gt;&amp;&gt;::__split_buffer(unsigned long, unsigned long, std::__1::allocator&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch&gt;&amp;) (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL+0x100127fde)
    #2 0x10e12ab1c in std::__1::__split_buffer&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch, std::__1::allocator&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch&gt;&amp;&gt;::__split_buffer(unsigned long, unsigned long, std::__1::allocator&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch&gt;&amp;) (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL+0x100126b1c)
    #3 0x10e128b86 in void std::__1::vector&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch, std::__1::allocator&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch&gt; &gt;::__push_back_slow_path&lt;OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch const&gt;(OpenMS::OPXLDataStructs::CrossLinkSpectrumMatch const&amp;) (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL+0x100124b86)
    #4 0x10e03b9e1 in TOPPOpenPepXL::main_(int, char const**) (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL+0x1000379e1)
    #5 0x1122c06e9 in OpenMS::TOPPBase::main(int, char const**) (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/lib/libOpenMS.dylib+0x40fc6e9)
    #6 0x10e007d87 in main (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL+0x100003d87)
    #7 0x7fff950f75ac in start (/usr/lib/system/libdyld.dylib+0x35ac)

HINT: if you don&apos;t care about these errors you may set ASAN_OPTIONS=detect_container_overflow=0.
If you suspect a false positive see also: https://github.com/google/sanitizers/wiki/AddressSanitizerContainerOverflow.
SUMMARY: AddressSanitizer: container-overflow (/Users/builder/jenkins/ws/openms_nightly_supplement/a2d943f6/build/bin/OpenPepXL+0x10003a0e1) in TOPPOpenPepXL::main_(int, char const**)
Shadow bytes around the buggy address:
  0x1c280001ab80: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x1c280001ab90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c280001aba0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c280001abb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa fa
  0x1c280001abc0: fa fa fa fa fa fa fa fa fc fc fc fc fc fc fc fc
=&gt;0x1c280001abd0: fc fc fc fc fc fc fc fc fc fc fc fc fc[fc]fc fc
  0x1c280001abe0: fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc fc
  0x1c280001abf0: fc fc fc fc fc fc fc fc fc fc fc fc fc fc fa fa
  0x1c280001ac00: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
  0x1c280001ac10: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c280001ac20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1844==ABORTING
</Value>
			</Measurement>
		</Results>
	</Test>
```

**Comment by timosachsenberg** [comment]
@enetz any insights?

**Comment by enetz** [comment]
Sorry for the delayed response.
I may have found the responsible code, but what would be the fastest way to test a fix?

**Comment by jpfeuffer** [comment]
No problem. You could try an ASan build by yourself (cmake (...) -DADDRESS_SANITIZER=On ../source) and run the tests. What platform are you on?
I can also build your branch on Jenkins if you comment it here.

**Comment by enetz** [comment]
I built an ASan build with clang 3.8.1 on Ubuntu, overall 776 tests failing.
OpenPepXL has a lot of memory leaks, but those are listed after the tool runs through all 25 spectra of the test without issues.
The error message is pointing to a `vector< OPXLDataStructs::CrossLinkSpectrumMatch >` object I think,
so I improved some unnecessarily abstruse code from before I coded the comparators for my data structures. If the error is not a false positive, than this should be the most likely place.
The fix is on my repo:
https://github.com/enetz/OpenMS/tree/fix/OPXL_ASan_error
